### DISCUSSION TITLE
Why is react-dom not part of peerDeps?

### DISCUSSION THREAD

**Discussion by AndrewPrifer** [jaredpalmer/tsdx#939] [General] [answered]
Just out of curiosity, why is react-dom not included in the peer dependencies in the react templates? It would make sense for me that they are included, and IIRC they used to be a couple versions ago.

**Comment by agilgur5** [comment, marked as answer]
> IIRC they used to be a couple versions ago.

`react-dom` was never a peerDep in the templates. React itself was only added as a peerDep in https://github.com/formium/tsdx/commit/9f5c6de9f8c7472fcd0dbbfe15aa18434e742d9f . The current `package.json` structure for the template code was created in #318, where [this line](https://github.com/formium/tsdx/pull/318/files#diff-74e7f08e6c7cd736e503a5c3549fcaeaab850f904c1ad2219f19ab0f9cd7530bR17-R19) left React as the only peerDep as well, and that line hasn't changed since.

> Just out of curiosity, why is react-dom not included in the peer dependencies in the react templates?

^That history would be the main reason. [CRL only has `react` as well](https://github.com/transitive-bullshit/create-react-library/blob/b4ad5b8f68d4b1253e161f3c1a78dc1bb7acea7f/template/typescript/package.json#L27).

I didn't write those templates, but another post-hoc rationalization may be: most React libraries generally do not import `react-dom`; `ReactDOM.render` is only called by the consumer of the library. React itself is meant to be generic with respect to renderer as well, and not all packages are built for DOM use or used by consumers in the DOM.

That being said, if your library has some limitation with respect to `react-dom` or is specifically built for DOM, it may be a good to list it as a peerDep, although it's an indirect relationship that may be a bit tenuous as a result.
The devDeps include `react-dom` as that is the most common use-case and is often used for testing, examples, Storybook, etc, so I can see a reason there for including it as peerDep as well for.

**Reply by AndrewPrifer** [reply]
I see, thank you for the thorough answer! I do in fact depend directly on react-dom, so it is an easy decision then.

**Reply by agilgur5** [reply]
Oh yea, if you actually depend directly on it, then for sure, it's better to have `react-dom` as a peerDep than a dep.

An example of the _indirect_ use can be found in one of my libraries, [`react-signature-canvas`](https://github.com/agilgur5/react-signature-canvas), where I have [a peerDep on `react-dom`](https://github.com/agilgur5/react-signature-canvas/blob/d449c18b5d3b7e3bda22bdafcc7e9ddba9e51cd7/package.json#L50). It doesn't actually import `react-dom` at all though, but it is very specific to DOM given that it uses `canvas`, `window`, etc.
Theoretically, someone could create an alternative DOM renderer to `react-dom` though and then a user could use either, so the peerDep could become a limitation in that case (but it hasn't yet -- if it became a problem, I'd expect a user to mention it in issues).

**Reply by AndrewPrifer** [reply]
That's a cool library!
